cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/RackAwareGoal.java

```diff
@@ -235,11 +235,13 @@ protected void rebalanceForBroker(Broker broker,
       throws OptimizationFailureException {
     LOG.debug("balancing broker {}, optimized goals = {}", broker, optimizedGoals);
     Set<String> excludedTopics = optimizationOptions.excludedTopics();
+    boolean onlyMoveImmigrantReplicas = optimizationOptions.onlyMoveImmigrantReplicas();
     // Satisfy rack awareness requirement.
     SortedSet<Replica> replicas = new TreeSet<>(broker.replicas());
     for (Replica replica : replicas) {
       if ((broker.isAlive() && satisfiedRackAwareness(replica, clusterModel))
-          || shouldExclude(replica, excludedTopics)) {
+          || shouldExclude(replica, excludedTopics)
+          || (onlyMoveImmigrantReplicas && !replica.isImmigrant())) {
```

Comment:
If a cluster is already violating a hard goal (e.g. rack awareness here), will this not allow movement of replicas between alive brokers?

`OPTION-1`: Should hard goals still allow replica moves across alive brokers (but as usual, they should prioritize immigrant replicas)?

`OPTION-2`: Otherwise, can we move `(onlyMoveImmigrantReplicas && !replica.isImmigrant())` logic inside `shouldExclude(...)`?

Reply:
I prefer moving `(onlyMoveImmigrantReplicas && !replica.isImmigrant())` logic inside `shouldExclude(...)`.